sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/function/FunctionService.java

```diff
@@ -0,0 +1,117 @@
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+import io.stargate.bridge.proto.Schema;
+import io.stargate.sgv2.docsapi.api.exception.ErrorCode;
+import io.stargate.sgv2.docsapi.api.exception.ErrorCodeRuntimeException;
+import io.stargate.sgv2.docsapi.api.v2.model.dto.DocumentResponseWrapper;
+import io.stargate.sgv2.docsapi.service.ExecutionContext;
+import java.util.List;
+import java.util.Objects;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+/**
+ * Service that executes document build-in functions. Currently supported:
+ *
+ * <ol>
+ *   <li><code>$pop</code>, implemented in {@link
+ *       io.stargate.sgv2.docsapi.service.function.impl.PopFunction}
+ *   <li><code>$push</code>, implemented in {@link
+ *       io.stargate.sgv2.docsapi.service.function.impl.PushFunction}
+ * </ol>
+ *
+ * @see BuiltInFunction
+ */
+@ApplicationScoped
+public class FunctionService {
+
+  // NOTE: Fully tested via the BuiltInFunctionResourceIntegrationTest
+
+  @Inject Instance<BuiltInFunction> functions;
+
+  /**
+   * Executes a built-in function.
+   *
+   * <p>Emits a failure in case:
+   *
+   * <ol>
+   *   <li>Function operation does not exists, with {@link
+   *       ErrorCode#DOCS_API_INVALID_BUILTIN_FUNCTION}
+   *   <li>Function execution failed, with any code
+   * </ol>
+   *
+   * @param namespace Namespace to execute the function for
+   * @param collection Collection to execute the function for
+   * @param documentId Document id to execute the function for
+   * @param subPath Document sub-path
+   * @param operation The function operation name.
+   * @param input Function input (optional, depends on the function).
+   * @param context Execution content
+   * @return Result of function invocation (depends on the function).
+   */
+  public Uni<DocumentResponseWrapper<JsonNode>> executeBuiltInFunction(
+      Uni<Schema.CqlTable> table,
+      String namespace,
+      String collection,
+      String documentId,
+      List<String> subPath,
+      String operation,
+      JsonNode input,
+      ExecutionContext context) {
+
+    // get the function
+    return findFunction(operation)
+
+        // if there, construct the data and call
+        .flatMap(
+            f -> {
+              BuiltInFunction.Data data =
+                  new BuiltInFunction.Data(
+                      namespace, collection, documentId, subPath, context, input);
+              return f.execute(table, data);
+            })
+
+        // if we get the result back, transform to the wrapper
+        .onItem()
+        .ifNotNull()
+        .transform(
+            result -> new DocumentResponseWrapper<>(documentId, null, result, context.toProfile()));
+  }
+
+  // finds a function, fails if not available
+  private Uni<BuiltInFunction> findFunction(String function) {
+    return Multi.createFrom()
+        .iterable(functions)
```

Comment:
Not a big deal with just 2 functions but is there no way to just build a simple lookup `Map`, keyed by name? Instead of iterating over instances.
Mostly asking wrt usage patterns, literally makes no measurable performance difference here :)

Reply:
Nope, because I need to Inject those instances.. This way adding a new function is just writing a new impl and adding to the context.. The benefit of this solution is that functions can also Inject anything they want..